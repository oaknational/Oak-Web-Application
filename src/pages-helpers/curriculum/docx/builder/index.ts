export { default as frontCover } from "./1_frontCover";
export { default as tableOfContents } from "./2_tableOfContents";
export { default as ourCurriculum } from "./3_ourCurriculum";
export { default as threadsExplainer } from "./4_threadsExplainer";
export { default as subjectExplainer } from "./5_subjectExplainer";
export { default as subjectPrincipals } from "./6_subjectPrincipals";
export { default as ourPartner } from "./7_ourPartner";
export { default as units } from "./8_units";
export { default as threadsOverview } from "./9_threadsOverview";
export { default as threadsDetail } from "./10_threadsDetail";
export { default as backCover } from "./11_backCover";
export { default as pageLayout } from "./12_pageLayout";
