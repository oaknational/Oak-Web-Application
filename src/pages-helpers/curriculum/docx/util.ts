import {
  Thread,
  Unit,
} from "@/components/CurriculumComponents/CurriculumVisualiser";
import { CombinedCurriculumData } from "@/pages/teachers/curriculum/docx-poc/[...slugs]";

export type UnitLike =
  | CombinedCurriculumData["units"][number]["unit_options"][number]
  | Unit;

export function notUndefined<TValue>(
  value: TValue | undefined,
): value is TValue {
  return value !== undefined;
}

// Safe version of String#includes(...)
export const textIncludes = (str: unknown, matchText: string) => {
  if (typeof str === "string") {
    return str.includes(matchText);
  }
  return false;
};

// Safe version of String#replace(...)
export const textReplacer = (
  input: unknown,
  selector: string,
  value: string,
): string => {
  if (typeof input === "string") {
    return input.replace(selector, value);
  }
  return input as string;
};

export function threadUnitByYear(units: Unit[], threadSlug: string) {
  const output = {} as Record<string, Unit[]>;

  units.forEach((unit: Unit) => {
    unit.threads.forEach((thread) => {
      if (thread.slug === threadSlug) {
        output[unit.year] = output[unit.year] ?? [];
        if (
          output[unit.year] &&
          // Check if unit is not already within output
          !output[unit.year]!.find((yearUnit) => yearUnit.slug === unit.slug)
        ) {
          output[unit.year]!.push(unit);
        }
      }
    });
  });

  return output;
}

// TODO: This is from from Sonali's work that's not yet merged
export function createThreadOptions(units: Unit[]): Thread[] {
  const threadOptions = [] as Thread[];

  units.forEach((unit: Unit) => {
    // Populate threads object

    unit.threads.forEach((thread) => {
      if (threadOptions.every((to) => to.slug !== thread.slug)) {
        threadOptions.push(thread);
      }
    });
  });

  // Sort threads

  const threadOrders = new Set(threadOptions.map((to) => to.order));
  if (threadOptions.length > threadOrders.size) {
    // In secondary science multiple threads can have the same order value due
    // to multiple subjects (eg biology, chemistry, physics) being shown, so
    // if orders are not unique, sort alphabetically by slug

    threadOptions.sort((a, b) => a.slug.localeCompare(b.slug));
  } else {
    // If orders are unique, use them to sort

    threadOptions.sort((a, b) => a.order - b.order);
  }
  return threadOptions;
}
