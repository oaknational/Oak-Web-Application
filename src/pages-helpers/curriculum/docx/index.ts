import { join } from "path";

import { format } from "date-fns";

import type {
  CurriculumOverviewMVData,
  CurriculumUnitsTabData,
} from "../../../node-lib/curriculum-api-2023";
import type { CurriculumOverviewSanityData } from "../../../common-lib/cms-types";

import * as builder from "./builder";
import {
  cmToEmu,
  cmToTwip,
  createImage,
  generateEmptyDocx,
  insertFooters,
  insertImages,
} from "./docx";

import { Ks4Option } from "@/node-lib/curriculum-api-2023/queries/curriculumPhaseOptions/curriculumPhaseOptions.schema";

export type CombinedCurriculumData = CurriculumUnitsTabData &
  CurriculumOverviewMVData &
  CurriculumOverviewSanityData;

export type Slugs = {
  subjectSlug: string;
  phaseSlug?: string;
  ks4OptionSlug?: string;
  keyStageSlug?: string;
  tierSlug?: string;
  childSubjectSlug?: string;
};

const ENABLE_TIMER = false;
const measure = (key: string, fn: () => void | Promise<void>) => {
  if (!ENABLE_TIMER) {
    const v = fn();
    return v;
  }
  let err: unknown;
  let v;
  console.time(key);
  try {
    v = fn();
  } catch (_err: unknown) {
    err = _err;
  } finally {
    console.timeEnd(key);
  }

  if (err) throw err;
  return v;
};

export default async function docx(
  data: CombinedCurriculumData,
  slugs: Slugs,
  ks4Options: Ks4Option[],
) {
  const zip = await generateEmptyDocx();

  const images = await insertImages(
    zip,
    {
      footerImage: join(
        process.cwd(),
        "src/pages-helpers/curriculum/docx/builder/images/footer-logo.png",
      ),
    },
    "footer-default.xml",
  );

  const footers = await insertFooters(zip, {
    default: `
      <w:p>
        <w:pPr>
            <w:jc w:val="right"/>
        </w:pPr>
        <w:r>
            <w:rPr>
                  <w:rFonts
                    w:ascii="Arial"
                    w:eastAsia="Arial"
                    w:hAnsi="Arial"
                    w:cs="Arial"
                  />
                  <w:rtl w:val="0"/>
            </w:rPr>
            <w:t xml:space="preserve">Exported ${format(
              new Date(),
              "dd MMMM yyyy",
            )}     </w:t>
            <w:tab/>
        </w:r>
        <w:r>
            <w:rPr>
              <w:rFonts
                w:ascii="Arial"
                w:eastAsia="Arial"
                w:hAnsi="Arial"
                w:cs="Arial"
              />
            </w:rPr>
            <w:fldChar w:fldCharType="begin"/>
            <w:instrText xml:space="preserve">PAGE</w:instrText>
            <w:fldChar w:fldCharType="separate"/>
            <w:fldChar w:fldCharType="end"/>
        </w:r>
        <w:r>
            <w:rPr>
                <w:rtl w:val="0"/>
            </w:rPr>
        </w:r>
        <w:r>
            ${createImage(images.footerImage, {
              width: cmToEmu(5.33),
              height: cmToEmu(0.55),
              xPos: cmToEmu(0.001),
              yPos: cmToEmu(0.001),
              xPosAnchor: "margin",
              yPosAnchor: "line",
              isDecorative: true,
            })}
        </w:r>
    </w:p>
    `,
  });

  // Run through the builders
  const runners = {
    frontCover: async () => await builder.frontCover(zip, { data, slugs }),
    frontCoverPageLayout: async () =>
      await builder.pageLayout(zip, {
        margins: {
          top: cmToTwip(1.5),
          right: cmToTwip(1.5),
          bottom: cmToTwip(1.5),
          left: cmToTwip(1.5),
          header: cmToTwip(1.5),
          footer: cmToTwip(1.5),
        },
      }),
    tableOfContents: async () =>
      await builder.tableOfContents(zip, { data, ks4Options }),
    tableOfContentsPageLayout: async () =>
      await builder.pageLayout(zip, {
        footers,
        margins: {
          top: cmToTwip(1.25),
          right: cmToTwip(1.25),
          bottom: cmToTwip(2.5),
          left: cmToTwip(1.25),
          header: cmToTwip(1.5),
          footer: cmToTwip(1.5),
        },
      }),
    ourCurriculum: async () => await builder.ourCurriculum(zip),
    threadsExplainer: async () =>
      await builder.threadsExplainer(zip, { slugs, data }),
    subjectExplainer: async () => await builder.subjectExplainer(zip, { data }),
    subjectPrincipals: async () => await builder.subjectPrincipals(),
    ourPartner: async () => await builder.ourPartner(zip, { data }),
    units: async () => await builder.units(zip, { data, slugs, ks4Options }),
    threadsOverview: async () => await builder.threadsOverview(zip, { data }),
    threadsOverviewPageLayout: async () =>
      await builder.pageLayout(zip, {
        footers,
        margins: {
          top: cmToTwip(1.5),
          right: cmToTwip(1.5),
          bottom: cmToTwip(2.5),
          left: cmToTwip(1.5),
          header: cmToTwip(1.5),
          footer: cmToTwip(1.5),
        },
      }),
    threadsDetail: async () => await builder.threadsDetail(zip, { data }),
    threadsDetailPageLayout: async () =>
      await builder.pageLayout(zip, {
        footers,
        margins: {
          top: cmToTwip(1.75),
          right: cmToTwip(1.75),
          bottom: cmToTwip(2.5),
          left: cmToTwip(1.75),
          header: cmToTwip(1.5),
          footer: cmToTwip(1.5),
        },
      }),
    backCover: async () => await builder.backCover(zip, { data }),
    backCoverPageLayout: async () =>
      await builder.pageLayout(zip, {
        footers,
        isLast: true,
        margins: {
          top: cmToTwip(1.25),
          right: cmToTwip(1.25),
          bottom: cmToTwip(2.5),
          left: cmToTwip(1.25),
          header: cmToTwip(1.5),
          footer: cmToTwip(1.5),
        },
      }),
  };

  // Note: Just some code to messure some timings.
  await measure("total", async () => {
    for (const [key, runner] of Object.entries(runners)) {
      await measure(key, async () => {
        await runner();
      });
    }
  });

  return await zip.zipToBuffer();
}
