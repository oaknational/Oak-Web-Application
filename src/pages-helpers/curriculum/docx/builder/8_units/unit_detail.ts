import type { Element } from "xml-js";

import { cdata, safeXml } from "../../xml";
import { CombinedCurriculumData } from "../..";
import {
  cmToEmu,
  cmToTwip,
  createImage,
  insertLinks,
  insertNumbering,
  JSZipCached,
  wrapInLinkTo,
} from "../../docx";
import { createUnitsListingByYear } from "../../tab-helpers";

import { getYearGroupTitle } from "@/utils/curriculum/formatting";
import { getUnitFeatures } from "@/utils/curriculum/features";
import { createProgrammeSlug } from "@/utils/curriculum/slugs";
import { SubjectCategory } from "@/utils/curriculum/types";
import { ENABLE_NEW_CURRIC_MV } from "@/utils/curriculum/constants";

const DISABLE_COLUMN_BREAKS = true;

type Unit = CombinedCurriculumData["units"][number];

async function buildUnitLessons(
  zip: JSZipCached,
  unit: Unit | Unit["unit_options"][number],
) {
  const numbering = await insertNumbering(zip, {
    lessonNumbering: safeXml`
      <XML_FRAGMENT>
        <w:lvl w:ilvl="0">
          <w:start w:val="1" />
          <w:numFmt w:val="decimal" />
          <w:lvlText w:val="%1." />
          <w:lvlJc w:val="left" />
          <w:pPr>
            <w:ind w:left="360" w:hanging="360" />
          </w:pPr>
          <w:rPr>
            <w:rFonts w:ascii="Arial" w:hAnsi="Arial" w:cs="Arial" />
          </w:rPr>
        </w:lvl>
      </XML_FRAGMENT>
    `,
  });

  const lessonsXmls =
    unit.lessons?.map((lesson) => {
      return safeXml`
        <w:p>
          <w:pPr>
            <w:numPr>
              <w:ilvl w:val="0" />
              <w:numId w:val="${numbering.lessonNumbering}" />
            </w:numPr>
            <w:spacing w:line="276" w:lineRule="auto" />
            <w:ind w:left="425" w:right="-17" w:hanging="360" />
          </w:pPr>
          <w:r>
            <w:rPr>
              <w:rFonts
                w:ascii="Arial"
                w:eastAsia="Arial"
                w:hAnsi="Arial"
                w:cs="Arial"
              />
              <w:color w:val="222222" />
            </w:rPr>
            <w:t>${cdata(lesson.title)}</w:t>
          </w:r>
        </w:p>
      ` as Element;
    }) ?? [];

  return lessonsXmls.join("");
}

async function buildUnitThreads(zip: JSZipCached, unit: Unit) {
  const numbering = await insertNumbering(zip, {
    threadsNumbering: safeXml`
      <XML_FRAGMENT>
        <w:nsid w:val="1E776843" />
        <w:multiLevelType w:val="multilevel" />
        <w:tmpl w:val="20BC537A" />
        <w:lvl w:ilvl="0">
          <w:start w:val="1" />
          <w:numFmt w:val="bullet" />
          <w:lvlText w:val="ï‚·" />
          <w:lvlJc w:val="left" />
          <w:pPr>
            <w:ind w:left="360" w:hanging="360" />
          </w:pPr>
          <w:rPr>
            <w:rFonts w:ascii="Symbol" w:hAnsi="Symbol" w:hint="default" />
          </w:rPr>
        </w:lvl>
      </XML_FRAGMENT>
    `,
  });

  if (unit.threads.length === 0) {
    return safeXml`
      <w:p>
        <w:r>
          <w:t>No threads</w:t>
        </w:r>
      </w:p>
    `;
  }

  const threadsXmls = unit.threads.map((thread) => {
    return safeXml`
      <w:p>
        <w:pPr>
          <w:numPr>
            <w:ilvl w:val="0" />
            <w:numId w:val="${numbering.threadsNumbering}" />
          </w:numPr>
          <w:spacing w:line="276" w:lineRule="auto" />
          <w:ind w:left="425" w:right="-17" w:hanging="360" />
        </w:pPr>
        <w:r>
          <w:rPr>
            <w:rFonts
              w:ascii="Arial"
              w:eastAsia="Arial"
              w:hAnsi="Arial"
              w:cs="Arial"
            />
            <w:color w:val="222222" />
          </w:rPr>
          <w:t>${cdata(thread.title)}</w:t>
        </w:r>
      </w:p>
    ` as Element;
  });
  return threadsXmls.join("");
}

function buildUnitOptionTitle(
  unitOption: CombinedCurriculumData["units"][number]["unit_options"][number],
  unitOptionIndex: number,
  images: { greenUnderline: string },
) {
  return safeXml`
    <XML_FRAGMENT>
      <w:p>
        <w:r>
          <w:rPr>
            <w:rFonts
              w:ascii="Arial"
              w:eastAsia="Arial"
              w:hAnsi="Arial"
              w:cs="Arial"
            />
            <w:b />
            <w:color w:val="222222" />
            <w:sz w:val="32" />
          </w:rPr>
          <w:t xml:space="preserve">Option ${cdata(unitOptionIndex + 1)}: </w:t>
          ${createImage(images.greenUnderline, {
            width: cmToEmu(2.74),
            height: cmToEmu(0.27),
            xPos: cmToEmu(-0.21),
            yPos: cmToEmu(0.7),
            xPosAnchor: "column",
            yPosAnchor: "paragraph",
            isDecorative: true,
          })}
        </w:r>
        <w:r>
          <w:rPr>
            <w:rFonts
              w:ascii="Arial"
              w:eastAsia="Arial"
              w:hAnsi="Arial"
              w:cs="Arial"
            />
            <w:color w:val="222222" />
            <w:sz w:val="32" />
          </w:rPr>
          <w:t>${cdata(unitOption.title)}</w:t>
        </w:r>
      </w:p>
      <w:p>
        <w:r>
          <w:t />
        </w:r>
      </w:p>
    </XML_FRAGMENT>
  `;
}

export async function buildUnit(
  zip: JSZipCached,
  unit: Unit,
  unitIndex: number,
  unitOption: undefined | Unit["unit_options"][number] | null,
  unitOptionIndex: undefined | number,
  images: {
    greenCircle: string;
    underline: string;
    jumpOutArrow: string;
    greenUnderline: string;
  },
  slugs: {
    ks4OptionSlug?: string;
    tierSlug?: string;
  },
) {
  const unitOptionIfAvailable = unitOption ?? unit;
  const unitNumber = unitIndex + 1;
  const lessons = await buildUnitLessons(zip, unitOptionIfAvailable);
  const threads = await buildUnitThreads(zip, unit);

  const hasPublishedLessons = (unitOptionIfAvailable.lessons ?? []).some(
    (lesson) => lesson._state === "published",
  );

  let resolvedUnitSlug;
  if (ENABLE_NEW_CURRIC_MV) {
    resolvedUnitSlug = unitOption?.slug ?? unit.slug;
  } else {
    resolvedUnitSlug = unitOption?.unitvariant_id
      ? `${unit.slug}-${unitOption.unitvariant_id}`
      : unit.slug;
  }

  const links = await insertLinks(zip, {
    onlineResources: `https://www.thenational.academy/teachers/programmes/${createProgrammeSlug(
      unit,
      slugs.ks4OptionSlug,
      slugs.tierSlug,
    )}/units/${resolvedUnitSlug}/lessons`,
  });

  let unitDescriptions: string = "";
  if (unit.cycle === "1") {
    const priorUnitTitle = unitOptionIfAvailable.connection_prior_unit_title
      ? unitOptionIfAvailable.connection_prior_unit_title
      : "-";
    const futureUnitTitle = unitOptionIfAvailable.connection_future_unit_title
      ? unitOptionIfAvailable.connection_future_unit_title
      : "-";
    unitDescriptions = safeXml`
      <XML_FRAGMENT>
        <w:p>
          <w:pPr>
            <w:pStyle w:val="Heading4" />
          </w:pPr>
          <w:r>
            <w:rPr>
              <w:rFonts
                w:ascii="Arial"
                w:eastAsia="Arial"
                w:hAnsi="Arial"
                w:cs="Arial"
              />
              <w:b />
              <w:i w:val="0" />
              <w:color w:val="000000" />
              <w:sz w:val="28" />
            </w:rPr>
            <w:t>Previous unit description</w:t>
          </w:r>
        </w:p>
        <w:p>
          <w:r>
            <w:rPr>
              <w:rFonts w:ascii="Arial" w:hAnsi="Arial" />
              <w:b />
            </w:rPr>
            <w:t>${cdata(priorUnitTitle)}</w:t>
          </w:r>
        </w:p>
        ${!unitOptionIfAvailable.connection_prior_unit_description
          ? ""
          : safeXml`
              <w:p>
                <w:r>
                  <w:rPr>
                    <w:rFonts w:ascii="Arial" w:hAnsi="Arial" />
                  </w:rPr>
                  <w:t>
                    ${cdata(
                      unitOptionIfAvailable.connection_prior_unit_description,
                    )}
                  </w:t>
                </w:r>
              </w:p>
            `}
        <w:p>
          <w:r>
            <w:rPr>
              <w:rFonts w:ascii="Arial" w:hAnsi="Arial" />
              <w:b />
            </w:rPr>
            <w:t />
          </w:r>
        </w:p>
        <w:p>
          <w:pPr>
            <w:pStyle w:val="Heading4" />
          </w:pPr>
          <w:r>
            <w:rPr>
              <w:rFonts
                w:ascii="Arial"
                w:eastAsia="Arial"
                w:hAnsi="Arial"
                w:cs="Arial"
              />
              <w:b />
              <w:i w:val="0" />
              <w:color w:val="000000" />
              <w:sz w:val="28" />
            </w:rPr>
            <w:t>Future unit description</w:t>
          </w:r>
        </w:p>
        <w:p>
          <w:pPr>
            <w:ind w:left="30" w:hanging="30" />
          </w:pPr>
          <w:r>
            <w:rPr>
              <w:rFonts
                w:ascii="Arial"
                w:eastAsia="Arial"
                w:hAnsi="Arial"
                w:cs="Arial"
              />
              <w:b />
              <w:color w:val="222222" />
              <w:sz w:val="24" />
            </w:rPr>
            <w:t>${cdata(futureUnitTitle)}</w:t>
          </w:r>
        </w:p>
        ${!unitOptionIfAvailable.connection_future_unit_description
          ? ""
          : safeXml`
              <w:p>
                <w:r>
                  <w:rPr>
                    <w:rFonts w:ascii="Arial" w:hAnsi="Arial" />
                  </w:rPr>
                  <w:t>
                    ${cdata(
                      unitOptionIfAvailable.connection_future_unit_description,
                    )}
                  </w:t>
                </w:r>
              </w:p>
            `}
      </XML_FRAGMENT>
    `;
  } else {
    const description = unitOptionIfAvailable.description
      ? unitOptionIfAvailable.description
      : "-";
    const whyThisWhyNow = unitOptionIfAvailable.why_this_why_now
      ? unitOptionIfAvailable.why_this_why_now
      : "-";
    unitDescriptions = safeXml`
      <XML_FRAGMENT>
        <w:p>
          <w:pPr>
            <w:pStyle w:val="Heading4" />
          </w:pPr>
          <w:r>
            <w:rPr>
              <w:rFonts
                w:ascii="Arial"
                w:eastAsia="Arial"
                w:hAnsi="Arial"
                w:cs="Arial"
              />
              <w:b />
              <w:i w:val="0" />
              <w:color w:val="000000" />
              <w:sz w:val="28" />
            </w:rPr>
            <w:t>Unit description</w:t>
          </w:r>
        </w:p>
        <w:p>
          <w:r>
            <w:rPr>
              <w:rFonts w:ascii="Arial" w:hAnsi="Arial" />
            </w:rPr>
            <w:t>${cdata(description)}</w:t>
          </w:r>
        </w:p>
        <w:p>
          <w:r>
            <w:rPr>
              <w:rFonts w:ascii="Arial" w:hAnsi="Arial" />
              <w:b />
            </w:rPr>
            <w:t />
          </w:r>
        </w:p>
        <w:p>
          <w:pPr>
            <w:pStyle w:val="Heading4" />
          </w:pPr>
          <w:r>
            <w:rPr>
              <w:rFonts
                w:ascii="Arial"
                w:eastAsia="Arial"
                w:hAnsi="Arial"
                w:cs="Arial"
              />
              <w:b />
              <w:i w:val="0" />
              <w:color w:val="000000" />
              <w:sz w:val="28" />
            </w:rPr>
            <w:t>Why this, why now?</w:t>
          </w:r>
        </w:p>
        <w:p>
          <w:pPr>
            <w:ind w:left="30" w:hanging="30" />
          </w:pPr>
          <w:r>
            <w:rPr>
              <w:rFonts
                w:ascii="Arial"
                w:eastAsia="Arial"
                w:hAnsi="Arial"
                w:cs="Arial"
              />
              <w:color w:val="222222" />
              <w:sz w:val="24" />
            </w:rPr>
            <w:t>${cdata(whyThisWhyNow)}</w:t>
          </w:r>
        </w:p>
      </XML_FRAGMENT>
    `;
  }

  function getSubjectCategoriesAsString(
    subjectcategories: SubjectCategory[] | null | undefined,
  ) {
    const combinedTitles = subjectcategories
      ? subjectcategories.map(({ title }) => title).join(", ")
      : "";
    return combinedTitles !== "" ? `: ${combinedTitles}` : "";
  }

  const yearData = createUnitsListingByYear([unit]);
  const yearTitle = getYearGroupTitle(
    yearData,
    getUnitFeatures(unit)?.programmes_fields_overrides?.year ?? unit.year,
  );

  const xml = safeXml`
    <XML_FRAGMENT>
      ${
        "" /*<w:p>
        <w:r>
          <w:rPr>
            <w:noProof />
          </w:rPr>
          ${createImage(images.greenCircle, {
            width: cmToEmu(1.77),
            height: cmToEmu(1.6),
            xPos: cmToEmu(1.55),
            yPos: cmToEmu(1.13),
            xPosAnchor: "page",
            yPosAnchor: "page",
            isDecorative: true,
          })}
        </w:r>
      </w:p>*/
      }
      <w:p>
        <w:pPr>
          <w:pStyle w:val="Heading3" />
        </w:pPr>
        <w:r>
          <w:rPr>
            <w:rFonts
              w:ascii="Arial"
              w:eastAsia="Arial"
              w:hAnsi="Arial"
              w:cs="Arial"
            />
            <w:b />
            <w:color w:val="222222" />
            <w:sz w:val="36" />
            <w:szCs w:val="36" />
          </w:rPr>
          <w:t>${cdata(unitNumber)}. ${cdata(unit.title)}</w:t>
        </w:r>
      </w:p>

      <w:p>
        <w:r>
          <w:rPr>
            <w:rFonts
              w:ascii="Arial"
              w:eastAsia="Arial"
              w:hAnsi="Arial"
              w:cs="Arial"
            />
            <w:b />
            <w:color w:val="222222" />
          </w:rPr>
          <w:t>${cdata(yearTitle)}</w:t>
          <w:t>
            ${cdata(getSubjectCategoriesAsString(unit.subjectcategories))}
          </w:t>
        </w:r>
      </w:p>

      <w:p />

      ${!hasPublishedLessons
        ? ""
        : safeXml`
            <w:p>
              ${wrapInLinkTo(
                links.onlineResources,
                safeXml`
                  <XML_FRAGMENT>
                    <w:r>
                      <w:rPr>
                        <w:rFonts
                          w:ascii="Arial"
                          w:eastAsia="Arial"
                          w:hAnsi="Arial"
                          w:cs="Arial"
                        />
                        <w:b />
                        <w:u w:val="single" />
                      </w:rPr>
                      <w:t>Go to unit resources</w:t>
                    </w:r>
                    <w:r>
                      <w:rPr>
                        <w:rFonts
                          w:ascii="Arial"
                          w:eastAsia="Arial"
                          w:hAnsi="Arial"
                          w:cs="Arial"
                        />
                        <w:b />
                        <w:u w:val="none" />
                      </w:rPr>
                      ${createImage(images.jumpOutArrow, {
                        width: cmToEmu(0.41),
                        height: cmToEmu(0.35),
                        isDecorative: true,
                      })}
                    </w:r>
                  </XML_FRAGMENT>
                `,
              )}
            </w:p>
          `}
      <w:p>
        <w:r>
          <w:rPr>
            <w:rFonts
              w:ascii="Arial"
              w:eastAsia="Arial"
              w:hAnsi="Arial"
              w:cs="Arial"
            />
          </w:rPr>
          <w:t />
        </w:r>
      </w:p>

      <w:p>
        <w:r>
          <w:rPr>
            <w:noProof />
            <w:color w:val="222222" />
          </w:rPr>
          ${createImage(images.underline, {
            width: cmToEmu(19.3),
            height: cmToEmu(0.16),
            xPos: cmToEmu(0.83),
            yPos: cmToEmu(0),
            yPosAnchor: "line",
            isDecorative: true,
          })}
        </w:r>
      </w:p>
      <w:p />
      ${unitOption && unitOptionIndex !== undefined
        ? buildUnitOptionTitle(unitOption, unitOptionIndex, images)
        : ""}
      <w:p>
        <w:pPr>
          <w:sectPr>
            <w:pgSz w:w="11909" w:h="16834" />
            <w:pgMar
              w:top="${cmToTwip(1.5)}"
              w:right="${cmToTwip(1.5)}"
              w:bottom="${cmToTwip(2.5)}"
              w:left="${cmToTwip(1.5)}"
              w:header="${cmToTwip(1.5)}"
              w:footer="${cmToTwip(1.5)}"
              w:gutter="0"
            />
            <w:cols w:space="720" />
          </w:sectPr>
        </w:pPr>
      </w:p>
      <w:p>
        <w:pPr>
          <w:pStyle w:val="Heading4" />
        </w:pPr>
        <w:r>
          <w:rPr>
            <w:rFonts
              w:ascii="Arial"
              w:eastAsia="Arial"
              w:hAnsi="Arial"
              w:cs="Arial"
            />
            <w:b />
            <w:i w:val="0" />
            <w:color w:val="000000" />
            <w:sz w:val="28" />
          </w:rPr>
          <w:t>Threads</w:t>
        </w:r>
      </w:p>
      ${threads}
      <w:p>
        <w:pPr>
          <w:ind w:left="30" w:hanging="30" />
          <w:rPr>
            <w:color w:val="222222" />
          </w:rPr>
        </w:pPr>
      </w:p>
      ${unitDescriptions}
      ${DISABLE_COLUMN_BREAKS ? safeXml` <w:p /> ` : ""}
      <w:p>
        <w:pPr>
          <w:pStyle w:val="Heading4" />
        </w:pPr>
        ${DISABLE_COLUMN_BREAKS
          ? ""
          : `<w:r>
          <w:br w:type="column" />
        </w:r>`}
        <w:r>
          <w:rPr>
            <w:rFonts
              w:ascii="Arial"
              w:eastAsia="Arial"
              w:hAnsi="Arial"
              w:cs="Arial"
            />
            <w:b />
            <w:i w:val="0" />
            <w:color w:val="000000" />
            <w:sz w:val="28" />
          </w:rPr>
          <w:t>Lessons in unit</w:t>
        </w:r>
      </w:p>
      ${lessons}
      <w:p>
        <w:pPr>
          <w:ind w:right="-1032" />
          <w:sectPr>
            <w:type w:val="continuous" />
            <w:pgSz w:w="11909" w:h="16834" />
            <w:pgMar
              w:top="${cmToTwip(1.5)}"
              w:right="${cmToTwip(1.5)}"
              w:bottom="${cmToTwip(2.5)}"
              w:left="${cmToTwip(1.5)}"
              w:header="${cmToTwip(1.5)}"
              w:footer="${cmToTwip(1.5)}"
              w:gutter="0"
            />
            <w:cols w:num="2" w:space="720" />
          </w:sectPr>
        </w:pPr>
      </w:p>
    </XML_FRAGMENT>
  `;

  return xml;
}
