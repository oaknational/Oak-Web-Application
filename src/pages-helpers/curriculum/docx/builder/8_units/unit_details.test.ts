import { JSZipCached } from "../../docx";

import { buildUnitPriorKnowledgeReqs } from "./unit_detail";

import { createUnit } from "@/fixtures/curriculum/unit";
import { ENABLE_PRIOR_KNOWLEDGE_REQUIREMENTS } from "@/utils/curriculum/constants";

const insertNumberingMock = jest.fn(() => {
  return {
    priorKnowledgeReqsNumbering: "test",
  };
});
jest.mock("../../docx", () => {
  return {
    ...jest.requireActual("../../docx"),
    insertNumbering: (...args: []) => insertNumberingMock(...args),
  };
});

const FAKE_ZIP = {};

describe("buildUnitPriorKnowledgeReqs", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  if (ENABLE_PRIOR_KNOWLEDGE_REQUIREMENTS) {
    test("with data", async () => {
      const out = await buildUnitPriorKnowledgeReqs(
        FAKE_ZIP as unknown as JSZipCached,
        createUnit({
          prior_knowledge_requirements: [
            "prior knowledge req 1",
            "prior knowledge req 2",
            "prior knowledge req 3",
          ],
        }),
      );
      expect(insertNumberingMock).toHaveBeenCalled();
      expect(insertNumberingMock).toHaveBeenCalledWith(
        FAKE_ZIP,
        expect.objectContaining({
          priorKnowledgeReqsNumbering: expect.anything(),
        }),
      );
      expect(out).toMatchSnapshot();
    });

    test("empty data", async () => {
      const out = await buildUnitPriorKnowledgeReqs(
        FAKE_ZIP as unknown as JSZipCached,
        createUnit({
          prior_knowledge_requirements: [],
        }),
      );
      expect(insertNumberingMock).not.toHaveBeenCalled();
      expect(out).toEqual(undefined);
    });

    test("no data", async () => {
      const out = await buildUnitPriorKnowledgeReqs(
        FAKE_ZIP as unknown as JSZipCached,
        createUnit({
          prior_knowledge_requirements: undefined,
        }),
      );
      expect(insertNumberingMock).not.toHaveBeenCalled();
      expect(out).toEqual(undefined);
    });
  }
});
