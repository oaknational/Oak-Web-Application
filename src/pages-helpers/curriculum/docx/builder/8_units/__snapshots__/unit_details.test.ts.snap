// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildUnitPriorKnowledgeReqs with data & !enabled 1`] = `undefined`;

exports[`buildUnitPriorKnowledgeReqs with data & enabled 1`] = `
"
      <w:p>
        <w:pPr>
          <w:pStyle w:val="Heading4" />
        </w:pPr>
        <w:r>
          <w:rPr>
            <w:rFonts
              w:ascii="Arial"
              w:eastAsia="Arial"
              w:hAnsi="Arial"
              w:cs="Arial"
            />
            <w:b />
            <w:i w:val="0" />
            <w:color w:val="000000" />
            <w:sz w:val="28" />
          </w:rPr>
          <w:t>Prior knowledge requirements</w:t>
        </w:r>
      </w:p>
      
          <w:p>
            <w:pPr>
              <w:numPr>
                <w:ilvl w:val="0" />
                <w:numId w:val="test" />
              </w:numPr>
              <w:spacing w:line="276" w:lineRule="auto" />
              <w:ind w:left="425" w:right="-17" w:hanging="360" />
            </w:pPr>
            <w:r>
              <w:rPr>
                <w:rFonts
                  w:ascii="Arial"
                  w:eastAsia="Arial"
                  w:hAnsi="Arial"
                  w:cs="Arial"
                />
                <w:color w:val="222222" />
              </w:rPr>
              <w:t><![CDATA[prior knowledge req 1]]></w:t>
            </w:r>
          </w:p>
        
          <w:p>
            <w:pPr>
              <w:numPr>
                <w:ilvl w:val="0" />
                <w:numId w:val="test" />
              </w:numPr>
              <w:spacing w:line="276" w:lineRule="auto" />
              <w:ind w:left="425" w:right="-17" w:hanging="360" />
            </w:pPr>
            <w:r>
              <w:rPr>
                <w:rFonts
                  w:ascii="Arial"
                  w:eastAsia="Arial"
                  w:hAnsi="Arial"
                  w:cs="Arial"
                />
                <w:color w:val="222222" />
              </w:rPr>
              <w:t><![CDATA[prior knowledge req 2]]></w:t>
            </w:r>
          </w:p>
        
          <w:p>
            <w:pPr>
              <w:numPr>
                <w:ilvl w:val="0" />
                <w:numId w:val="test" />
              </w:numPr>
              <w:spacing w:line="276" w:lineRule="auto" />
              <w:ind w:left="425" w:right="-17" w:hanging="360" />
            </w:pPr>
            <w:r>
              <w:rPr>
                <w:rFonts
                  w:ascii="Arial"
                  w:eastAsia="Arial"
                  w:hAnsi="Arial"
                  w:cs="Arial"
                />
                <w:color w:val="222222" />
              </w:rPr>
              <w:t><![CDATA[prior knowledge req 3]]></w:t>
            </w:r>
          </w:p>
        
    "
`;
